aten/src/ATen/native/sparse/SparseTensorMath.cpp

```diff
@@ -1321,4 +1322,157 @@ Tensor any_sparse(const Tensor& self) {
   return at::any(self._values());
 }
 
+Tensor bmm_sparse_cpu(const SparseTensor& self, const Tensor& mat2) {
+  Tensor result = at::empty({}, mat2.options());
+  return bmm_out_sparse_cpu(result, self, mat2);
+}
+
+// Search a sorted strided array for the rightmost instance of a value.
+// Array must be sorted from lowest to highest
+template<typename scalar_t>
+scalar_t binary_search_strided_rightmost(scalar_t search_val, scalar_t* sorted_arr, int64_t sorted_arr_stride, int64_t length, bool* found) {
+  if (length == 0) {
+    *found = false;
+    return -1;
+  }
+
+  int64_t left_ind = 0;
+  int64_t right_ind = length - 1;
+  int64_t mid_ind;
+  bool done_searching = false;
+
+  while (!done_searching) {
+    mid_ind = (left_ind+right_ind) >> 1;
+    scalar_t mid_val = sorted_arr[mid_ind*sorted_arr_stride];
+
+    if (mid_val > search_val) {
+      right_ind = mid_ind-1;
+    } else if((mid_val == search_val) && (
+      (mid_ind == length - 1) || (sorted_arr[(mid_ind+1)*sorted_arr_stride] != search_val)
+    )) {
+      done_searching = true;
+      *found = true;
+    } else {
+      left_ind = mid_ind+1;
+    }
+
+    if (left_ind > right_ind) {
+      done_searching = true;
+      *found = false;
+      mid_ind = -1;
+    }
+  }
+
+  return mid_ind;
+}
```

Comment:
Since it is the `rightmost` variant, would it be more efficient to return `right_ind`? The logic seems like if `right_ind` already hits the rightmost, the method will not return until it is hit from the left...

Reply:
I think you're right. The logic would be simpler that way. I'll change it and see how it affects performance.

Reply:
Actually, most of the time when this search function is called, if the number of non-zero elements per matrix is roughly equal, the index we're searching for is going to be somewhere on the very left side of the array. The caller of this function is searching for the matrix indices in order, and it's shrinking the search area each time by starting from the right of the previously-found position. So its probably more likely that mid_ind will hit the position we're looking for before right_ind does. And anyway, this search is usually going to be a very small portion of the total runtime of bmm. I tried a basic linear search and the run time difference was very small, only about 1% slower in most of my test cases.